lib/ansible/modules/cloud/vmware/vmware_guest_file_operation.py

```diff
@@ -0,0 +1,332 @@
+    vm_password = module.params['vm_password']
+
+    if "src" not in module.params["copy"]:
+        module.fail_json(msg="copy.src is mandatory")
+    if "dest" not in module.params["copy"]:
+        module.fail_json(msg="copy.dest is mandatory")
+    if "overwrite" in module.params["copy"]:
+        overwrite = module.params["copy"]["overwrite"]
+    else:
+        overwrite = False
+
+    dest = module.params["copy"]['dest']
+    src = module.params['copy']['src']
+    b_src = to_bytes(src, errors='surrogate_or_strict')
+    if not os.path.exists(b_src):
+        module.fail_json(msg="Source %s not found" % (src))
+    if not os.access(b_src, os.R_OK):
+        module.fail_json(msg="Source %s not readable" % (src))
+    if os.path.isdir(b_src):
+        module.fail_json(msg="copy does not support copy of directory: %s" % (src))
+
+    data = None
+    with open(b_src, "r") as local_file:
+        data = local_file.read()
+    file_size = os.path.getsize(b_src)
+
+    creds = vim.vm.guest.NamePasswordAuthentication(username=vm_username, password=vm_password)
+    file_attributes = vim.vm.guest.FileManager.FileAttributes()
+    file_manager = content.guestOperationsManager.fileManager
+    url = file_manager.InitiateFileTransferToGuest(vm=vm, auth=creds, guestFilePath=dest,
+                                                   fileAttributes=file_attributes, overwrite=overwrite,
+                                                   fileSize=file_size)
+
+    resp, info = urls.fetch_url(module, url, data=data, method="PUT")
+    status_code = info["status"]
+    if status_code != 200:
+        raise Exception('initiateFileTransferToGuest : problem during file transfer')
+    return True
+
+
+def main():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(dict(datacenter=dict(type='str'),
+                              cluster=dict(type='str'),
+                              folder=dict(type='str', default='/vm'),
+                              vm_id=dict(type='str', required=True),
+                              vm_id_type=dict(default='vm_name', type='str',
+                                              choices=['inventory_path', 'uuid', 'dns_name', 'vm_name']),
+                              vm_username=dict(type='str', required=True),
+                              vm_password=dict(type='str', no_log=True, required=True),
+                              directory=dict(type='dict', default={}),
+                              copy=dict(type='dict', default={}),
+                              fetch=dict(type='dict', default={})
+                              ))
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=False,
+                           required_if=[['vm_id_type', 'inventory_path', ['folder']]],
+                           )
+
+    if not HAS_PYVMOMI:
+        module.fail_json(changed=False, msg='pyvmomi is required for this module')
+
```

Comment:
add a check 

if vm_id_type is vm_name and not folder then present error "Folder is required parameter when vm_id_type is inventory_path"